File: proj05/ruby/quadRoot.rb
#! /usr/bin/ruby
# quadRoot.rb is a driver for computeRoots.
#
# Author: David Reidsma
# Date:   03/11/2020

# computeRoots computes the roots of a quadratic equation.
# Input: a, a float
# Input: b, a float
# Input: c, a float
# Output: A tuple storing a boolean (true for success, false otherwise) and two numbers (the roots)
def computeRoots(a, b, c)
  if a != 0
    arg = (b ** 2.0) - (4 * a * c)
    if arg >= 0
      root1 = (-b + Math.sqrt(arg)) / (2 * a);
      root2 = (-b - Math.sqrt(arg)) / (2 * a);
      return true, root1, root2
    else
      print "b^2 - 4ac is negative!"
      return false, 0.0, 0.0
    end
  else
    print "a is zero!"
    return false, 0.0, 0.0
  end
end

# Main function of a program that computes quadratic roots
if __FILE__ == $0
  puts "Enter a: "
  a = gets.chomp.to_f
  puts "Enter b: " 
  b = gets.chomp.to_f
  puts "Enter c: "
  c = gets.chomp.to_f

  success, root1, root2 = computeRoots(a, b, c)
  if success
    print "#{root1}, #{root2}"
  end
end

File: proj05/clojure/src/quad_root.clj
;;;; quad_root.clj is a driver for computeRoots.
;;;;
;;;; Author:    David Reidsma
;;;; Date:      03/11/2020

(ns quad_root)

;;; computeRoots computes the roots of a quadratic equation.
;;; Input: a, a float
;;; Input: b, a float
;;; Input: c, a float
;;; Output: A vector storing a boolean (true for success, false otherwise) and two numbers (the roots)
(defn computeRoots [a, b, c]
  (if (not (zero? a))
    (let
      [arg (- (Math/pow b 2) (* 4 a c))]
      (if (>= arg 0)
        (let [root1 (/ (+ (- b) (Math/sqrt arg)) (* 2 a))
              root2 (/ (- (- b) (Math/sqrt arg)) (* 2 a))]
          (vector true root1 root2)
        )
        (do
          (println "b^2 - 4ac is negative!")
          (vector false 0 0)
        )
      )
    )
    (do
      (println "a is zero!")
      (vector false 0 0)
    )
  )
)

;;; Main function of a program that computes quadratic roots
(defn -main []
  (print "Enter a: ") (flush)
  (let [a (read)]
    (print "Enter b: ") (flush)
    (let [b (read)]
      (print "Enter c: ") (flush)
      (let [c (read)
            roots (computeRoots a b c)]
        (if (get roots 0)
          (printf "%f, %f\n" (get roots 1), (get roots 2))
        )
      )
    )
  )
)
